### DISCUSSION TITLE
Third party auth login: "Unable to exchange external code: <code>"

### DISCUSSION THREAD

**Discussion by clairefro** [supabase/supabase#1192] [Questions] [answered]
I'm trying to implement third party auth with Github by passing `{ provider: 'github' }`. On signin attempt, I get a redirect with the following query params in the url: 

```
error: server_error
error_description: Unable to exchange external code: 7cf84a55c55bd5f9ac26
```

How can I resolve this error?

**Comment by PaulRennerbs** [comment]
I have the same issue

**Reply by awalias** [reply]
a lot of people are reporting getting spaces when they copy-paste the github client secret, might be worth checking

**Reply by TuanManhCao** [reply]
Thanks @awalias . This actually my issues. And this is the correct answer.

Can we have some UX improvement here? Like highlight "Space" character or error validation when space character is input?

**Comment by ElesGergo** [comment]
same issue

**Reply by awalias** [reply]
a few people are reporting getting spaces when they copy-paste the github client secret, might be worth checking

**Comment by kangmingtay** [comment]
Hi everyone, apologies for the late reply, we'll be investigating this issue here (https://github.com/supabase/gotrue/issues/106#issue-900194371) and we aim to get back as soon as possible!

**Reply by subhaschnadraBodaki** [reply]
i am getting the issue using azure

**Comment by awalias** [comment]
hey @ElesGergo , @PaulRennerbs , @clairefro could you please share your project ref (the one from your supabase url) with beta@supabase.io with a link to this discusson so we can debug! Thanks!

**Reply by awalias** [reply]
also there should be some API logs at the bottom of the Auth > Settings page if you could include anything relevant from there

**Comment by davitykale** [comment]
I'm getting the same error with Apple auth

**Reply by kangmingtay** [reply]
Hi @davitykale, could you please share the steps you took to enable Apple as a provider [here](https://github.com/supabase/gotrue/issues/106)? This error usually appears when the resource provider fails to convert the authorization code into a token after the user is redirected back to the callback URL.

**Reply by jonathan-pyt** [reply]
Hi, how did you fix it? At first it worked for 6 months, and now my users can't sign in with apple (didn't touch anything).. i got the exact same error message and i don't know how to fix it :/ maybe i should restart the steps to add apple sign in ?

**Reply by kangmingtay** [reply]
Hey @Mistes974, you can email us at support@supabase.com and we'll take a look at your project.

**Reply by jonathan-pyt** [reply]
Thanks to @kangmingtay support.. i had to regenerate apple client solve the issue ! 
Because of this  `validity_period = 180 # In days. Max 180 (6 months) according to Apple docs.`

**Reply by mikob** [reply]
@Mistes974 @kangmingtay I'm running into the same issue. I created my keys 6 months ago. I tried regenerating my apple client secret by running `ruby secret_gen.rb` and putting the outputted secret into the apple secret section here: https://app.supabase.io/project/[project-id]/auth/settings and saving. Still getting the same error. Was there anything else you needed to do? Can the instructions for renewing the key be included in these docs since it needs to be done relatively frequently? https://supabase.com/docs/guides/auth/auth-apple

**Reply by farzd** [reply]
this worked for me,
i made a new key here: https://developer.apple.com/account/resources/authkeys/list
updated the key_file and key_id props inside secret_gen.rb
generated a new client_secret.txt with secret_gen.rb
and updated the client secret for apple sign in inside supabase auth settings [https://app.supabase.io/project/xxxx/auth/settings]

**Comment by kangmingtay** [comment, marked as answer]
@clairefro @PaulRennerbs @ElesGergo @davitykale I did some digging around and managed to reproduce it on my end by forcefully changing the `redirect_uri` in the body of the POST request when the server attempts to convert the authorization code into a token. It might be worthwhile to check if the `redirect_uri` is modified or stripped by an external proxy. The `redirect_uri` should match the redirect_url / callback_url you whitelisted in the resource provider app.

**Reply by seenickcode** [reply]
It seems that this is still an issue, at least when I use http://localhost:3000 as the redirect URL via my provider settings (Github). I set `supabase.auth.signIn({ redirectTo: 'http://localhost:3000' })` as well but no luck. Any further ideas?

**Reply by kangmingtay** [reply]
> It seems that this is still an issue, at least when I use http://localhost:3000 as the redirect URL via my provider settings (Github). I set `supabase.auth.signIn({ redirectTo: 'http://localhost:3000' })` as well but no luck. Any further ideas?

Hey @seenickcode,if you're trying to implement sign-up with an OAuth provider, I believe the endpoint in the docs is incorrect. It should be `supabase.auth.signUp({ provider: 'github' }, { redirectTo: 'http://localhost:3000' })` Else, could you please provide the error message you encountered? Hope this helps!

**Reply by seenickcode** [reply]
Hi Kang, thanks for your help here. By endpoint do you mean function call? The example you cited doesn't work as it should be `signIn`, not `signUp` if using a provider. Also, I have already tried `redirectTo` option with no luck.

**Reply by kangmingtay** [reply]
Hi @seenickcode, my bad, yeah I meant the function call. I just tried doing this and it works for me:
```
let { error } = await supabase.auth.signIn({ 'github' }); 
```
Can you please kindly double check that you have the correct callback url entered in your github oauth application settings? It should be in the form of `https://<PROJECT_REF>.supabase.co/auth/v1/callback`

As for the `redirectTo`, we are aware of this issue and we are tracking it in our gotrue repo [here](https://github.com/supabase/gotrue/issues/142).

If this doesn't help, can you provide the error message you encountered? Thanks

**Reply by seenickcode** [reply]
So note, this _was_ working yesterday and I'm not sure what went wrong here. 

1. I switched to your code example, `let { error } = await supabase.auth.signIn({ 'github' });`

2.  I've confirmed I'm using the right Github client ID and secret in my supabase Settings for this Provider.

3.  I've confirmed my redirect URL for my Github app is set to: `https://eplluurobeymmdnfyxpy.supabase.co/auth/v1/callback`

4. When I sign in with Github, it redirects me back to this URL `http://localhost:3000/?error=server_error&error_description=Unable+to+exchange+external+code%3A+73d3812f6e23137cbacf` I also do not get any user, session or even any error object back from supabase, they are all null.

Additional notes:

- This works fine with the Google provider.
- Works fine with email signup.

Thanks for your help thus far.

**Reply by kangmingtay** [reply]
Hey @seenickcode, this problem could exist because of the following reasons:
1. An invalid client or secret key was entered in the Auth settings page. (check for any leading or ending spaces) 
2. An error occured on github's token endpoint

Since it is unlikely for 2 to occur, I highly recommend checking your client and secret token fields again (make sure there are no strange characters or spaces). I've made that mistake once when copying & pasting which led to this error too 🙃 

If it doesn't solve your problem, let me know and i'll restart the gotrue service for your project and see if it helps.

**Reply by seenickcode** [reply]
Ok. Note it did used to work. Maybe I will try it again with a new client ID and secret. Will let you know if you need to restart. Thanks for your help.

**Reply by seenickcode** [reply]
Ok, turns out there was a _leading_ space in the secret I had pasted into the Supabase Auth section for github client secret. Works now. Thank you for your help!

**Comment by bramaudi** [comment]
Hello, i face the same issue here, but the problem is gone when i removing space in github secret client (caused by mistake when copy & paste).

**Reply by kangmingtay** [reply]
It's always those pesky leading / trailing spaces that get you! On a side note, we've opened an issue [here](https://github.com/supabase/gotrue/issues/158) to preprocess the OAuth credentials to reduce the degree of human error :)

**Comment by solvemon** [comment]
Apologies for resurrecting an old issue, but I would like to note that I encountered this same issue today.

I have three running Supabase instances (2 hosted for prod and test, and 1 locally hosted for dev), but only the local one had this problem. The three instances are set up using the same Azure app-registration - so not a problem with key expiry. 

I only suddenly got this problem today (very randomly it seems) on my local hosted instance. Since there was no change in client key or secret since yesterday, I tried to restart the local supabase backend (`supabase stop && supabase start`), and it solved the problem!

**Reply by kangmingtay** [reply]
please open a ticket at https://app.supabase.com/support/new so we can investigate the issue

**Comment by razanfawwaz** [comment]
Hi, I'm using azure and got the same error. And there's no mistake when copy & paste the client id & secret id.

**Reply by kangmingtay** [reply]
please open a ticket at https://app.supabase.com/support/new so we can investigate the issue

**Reply by razanfawwaz** [reply]
I have already solved it! In the dashboard, I put the secret id from Azure, but the right is the value from Azure! not the secret id.
<img width="802" alt="image" src="https://user-images.githubusercontent.com/40789105/180719957-6f6964aa-2a62-4242-b1fa-7403f9d05475.png">
Maybe we can change the `Secret ID` to the `Secret Value`. Cause I think it will help others easier and more understandable when reading the docs.

**Reply by razanfawwaz** [reply]
@kangmingtay I make a pull request #7882 that maybe you can check. I add the highlight in value and add the documentation to using Azure Tenant URL

**Reply by kangmingtay** [reply]
thanks for the contribution @razanfawwaz !! :)

**Reply by Erildo** [reply]
@kangmingtay  I'm having a problem with Supabase using linkedin as a provider, When I login with my profile from linkedin it works fine and to test it further I created a new account in linkedin and tried to login the same way,it throws an error:

Unable to exchange external code

**Reply by kangmingtay** [reply]
hey @Erildo, please contact us at support with your project reference id so we can investigate further.

**Comment by 0xJoyce** [comment]
I am having same issue but with Google auth via Supabase.

**Reply by kangmingtay** [reply]
please create a support ticket (https://app.supabase.com/support/new) so we can investigate further

**Reply by imansyed0** [reply]
hi same issue - did you find a solution for this with Google sign in?

**Comment by Neesh774** [comment]
Having the same issue with Figma Auth, tried creating a support ticket and they told me to ask here :/

**Reply by kangmingtay** [reply]
Hey @Neesh774, we've replied to your support ticket, but just for visibility:

> According to your project's auth logs (https://supabase.com/dashboard/project/_/logs/auth-logs), if you look at the errors in the /callback endpoint, you can see that the reason behind the "Unable to exchange external code" error is due to exceeding your rate limit quota on figma.

**Comment by AlexIsMaking** [comment]
I had this issue with Notion Auth because I hadn't given my Notion integration access to user's email addresses. The auth logs showed more detail about what was causing the issue.

**Comment by sahil-lalani** [comment]
having the same issue but for discord oauth. getting a similar but different error: "500: Error getting user email from external provider."

**Comment by wishup-agrit** [comment]
Hey, I was just exploring the supabase auth, and I am stuck at apple sign in for my react-native web build
```
http://localhost:8081/?error=server_error&error_code=500&error_description=Unable+to+exchange+external+code:+c315aa584b9844e098a3b44df9daacb97.0.rtss.1yYLLHi-jXGHt9lERmiZUg#error=server_error&error_code=500&error_description=Unable+to+exchange+external+code%253A+c315aa584b9844e098a3b44df9daacb97.0.rtss.1yYLLHi-jXGHt9lERmiZUg
```
I have checked for trailing space and followed every step

**Comment by enqvida** [comment]
In my case, when I copied the Apple Developer ID from the top corner of the website by following this tutorial https://www.youtube.com/watch?v=6I2JEky20ME, and pasted it into the Supabase tool to generate a new Apple client secret, the Apple Developer ID was not respecting capital letters. It took me quite a while to figure it out. Be careful!

**Comment by ioancruso** [comment]
I was experiencing the same thing, until I found that I put the GOOGLE_CLIENT_ID as GOOGLE_CLIENT_SECRET. My mistake!

**Comment by manoelfilho** [comment]
Hello, everybody
I am facing the same issue here using Azure provider. Always seeing the same error
**Unable to exchange external code: M.C54 ...**

this is the simple function I am trying to use. It is from the Supabase Doc:

`
func loginMicrosoft() async throws {
        try await SupabaseClientManager.shared.client.auth.signInWithOAuth(
            provider: .azure,
            redirectTo: URL("my-app://login")!,
            scopes: "email")
    }
`
    
    Is there somebody experiencing the same. I am using the supababe 2.5.1

**Reply by manoelfilho** [reply]
Ok. I think I did something wrong. I remade the App in the Azure account and used the option:
**Accounts in any organizational directory (Any Microsoft Entra ID tenant - Multitenant) 
and personal Microsoft accounts (e.g. Skype, Xbox)**
 in my free Azure account and it worked. I need to do tests before having a paid account on Azure, so I was using a free account with a company domain and even if you can use it, it seems that the authentication simply does not work. It only works (and well) with a free account for common emails.
 
 My code with Supabase 2.25:

```swift
func loginMicrosoft() async throws -> AuthUser {
        do {
            let session = try await Supabase.auth.signInWithOAuth(
                provider: .azure,
                redirectTo: URL(string: <redirectUrl>)!,
                scopes: "openid email profile",
                queryParams: [(name: "prompt", value: "login")]
            )
            let user: Supabase.User = session.user
            return AuthUser(id: user.id, email: user.email!, token: session.accessToken, name: "User Microsoft", deadlineToken: Date(timeIntervalSince1970: session.expiresAt))
        } catch {
            throw AuthError.wrongCredentials
        }
    }
```

**Comment by JungeWerther** [comment]
(edit:) are you sure you 3rd party client credentials are correct? Normally there should be a log in the Authentication logs in the dashboard that can provide more info

**Comment by Dev-Salem** [comment]
I faced the same issue with google sign in on web (i already have the android one configured). make sure to add all the client ids with comma separated. also for web the client id SHOULD be the first one. (e.g `web-client-id,android-client-id,`)

**Comment by oskar-makarov** [comment]
If you're landing here from search: when Apple sign-in worked fine and then suddenly started failing with `Unable to exchange external code` / `invalid_client` after ~6 months — your Apple client secret has expired. Unlike every other OAuth provider, Apple's "client secret" is a signed JWT with a maximum lifetime of 6 months (15777000 seconds). It expires silently — no warning from Apple, no error in the Supabase dashboard.

The quick fix is to generate a new secret from your `.p8` key and paste it into Authentication → Providers → Apple.

The permanent fix is to automate the rotation. The Supabase docs recommend setting a recurring calendar reminder, but I automated it instead with a GitHub Action that regenerates the JWT from your `.p8` and updates your project via the Management API on a schedule: https://github.com/oskar-makarov/apple-client-secret-rotator

One-time 5-minute setup (4 repo secrets + a cron workflow), then it rotates every 5 months on its own. Zero dependencies — the whole thing is one readable file. Disclosure: I'm the author; it's MIT-licensed and free.
